src/backend/distributed/transaction/transaction_management.c

```diff
@@ -806,6 +818,22 @@ PopSubXact(SubTransactionId subId)
 	 */
 	activeSetStmts = state->setLocalCmds;
 
+	/*
+	 * keep subtransaction's propagated objects at toplevel transaction
+	 * if subtransaction committed.
+	 */
+	if (commit)
+	{
+		HASH_SEQ_STATUS propagatedObjectsSeq;
+		hash_seq_init(&propagatedObjectsSeq, state->propagatedObjects);
+		ObjectAddress *objectAddress = NULL;
+		while ((objectAddress = hash_seq_search(&propagatedObjectsSeq)) != NULL)
+		{
+			hash_search(PropagatedObjectsInTx, objectAddress, HASH_ENTER, NULL);
+		}
+		hash_delete_all(state->propagatedObjects);
+	}
```

Comment:
From my understanding of this block and the rest of the subtransaction sub system is that sub transactions are nested.

My assumption is that if an inner sub transaction commits, but its outer subtransaction aborts the earlier commited subtransaction becomes void as well.

With that in mind the block above seems to be a bit strange. We always move object addresses from a subtransaction into the root transaction on commit. This means that non-existing object addresses can be in the root transaction. Although this should not cause problems as object addresses should not be reused.

It seems strange we distinguish between commit and abort in this code it there are still routes where non-committed object addresses could end up in the root transaction.

I would suggest to either;

1. document why it is ok to have non-exisiting object addresses in the root transaction and always add them there.
2. if we want to go the subtransaction route, move the object addresses to the parent subtransaction here, instead of to the root. The inner most subtransaction on the stack would finally need to move the addresses to the root.

Now there is a subtile inconsistency in what the datastructures mean, and what we intend to do, compared to what we are actually doing.